File: src/main/java/bruteforce/BJ14889.java
package bruteforce;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BJ14889 {
	static int n, ans;
	static int[][] arr;
	static int[] visit;
	
	public static void cal( int index, int depth ) {
		if( depth == (n/2) ) {
			int t1 = 0;
			int t2 = 0;
			
			for( int i=0; i<n; i++ ) {
				for( int j=i+1; j<n; j++ ) {
					if( visit[i] == 1 && visit[j] == 1 ) {
						t1 += arr[i][j];
						t1 += arr[j][i];
					} else if( visit[i] == 0 && visit[j] == 0 ) {
						t2 += arr[i][j];
						t2 += arr[j][i];
					}
				}
			}
			
			int cha = Math.abs( t1-t2 );
			if( cha == 0 ) {
				System.out.println( cha );
				System.exit( 0 );
			}
			
			ans = Math.min( ans, cha );
			return;
			
		}
		
		for( int i=0; i<n-1; i++ ) {
			if( visit[i] == 0 ) {
				visit[i] = 1;
				cal( i+1, depth+1 );
				visit[i] = 0;
			}
		}
		
		
	}

	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		n = Integer.parseInt( br.readLine() );
		arr = new int[n][n];
		visit = new int[n];
		StringTokenizer st;
		ans = Integer.MAX_VALUE;
		for( int i=0; i<n; i++ ) {
			st = new StringTokenizer( br.readLine() );
			for( int j=0; j<n; j++ ) {
				arr[i][j] = Integer.parseInt( st.nextToken() );
			}
		}
		
		cal( 0, 0 );
		System.out.println( ans );
		
	}
}

File: Python/bruteforce/BJ14889.py
from sys import stdin
def cal( index, depth ):
    global n, arr, visit, ans
    if depth == (n//2):
        t1 = 0; t2 = 0
        for i in range( n ):
            for j in range( i+1, n ):
                if visit[i] and visit[j]:
                    t1 += arr[i][j]
                    t1 += arr[j][i]
                elif visit[i]==False and visit[j]==False:
                    t2 += arr[i][j]
                    t2 += arr[j][i]
        
        cha = abs( t1-t2 )
        if cha == 0:
            print( cha )
            exit()
        
        ans = min( [ans, cha] )
        return
    
    for i in range( index, n-1 ):
        if not visit[i]:
            visit[i] = True
            cal( i+1, depth+1 )
            visit[i] = False

input = stdin.readline
n = int(input())
arr = []
for _ in range( n ):
    arr.append( list( map(int, input().split()) ) )

visit = [False] * n
ans = 9999
cal( 0, 0 )
print( ans )